### ISSUE TITLE
TIP when running big backtests: increase memory size

### ISSUE THREAD

**Issue by thegamecat** [askmike/gekko#929]
Node has a memory limit of 1.5gig. In version 0.5.0 something caused long backtests (e.g. 16 month 1m candles) to fail with memory exhaustion. 

Initially not using talib indicators seemed to be the fix. However, I've recently encountered the issue again when only using native indicators which is very annoying when running a GA over night...

Anyway, I am a noob, all I needed was to expand the memory. You do this by adding the flag listed below to the node command. Now, it seems all is good.

node --max-old-space-size=8192 server.js  

I believe it's worth including this in the documentation as a tip  (for those noobs like me anyway )

**Comment by askmike** [comment]
> caused long backtests (e.g. 16 month 1m candles) to fail with memory exhaustion.

Just to be sure: this only happens for strategies that mix native and talib indicators.

> I believe it's worth including this in the documentation as a tip (for those noobs like me anyway )

Yes sure! But you are definitely not a noob anymore ;)

**Comment by thegamecat** [comment]
Actually I encountered the memory issue in the following scenario:

Using the api (running node server.js) with 8 backtest requests. Each backtest uses 20-50MB and 10-14% cpu. The server.js thread (as shown in windows task manager) eats up memory (I am assuming this is total memory usage as each of the backtests builds their various candle / data arrays. Even without talib or a talib / native indicator mix, this causes the memory exhaustion. 

TLDR: any backtest with any type of indicators (native, talib or a mix) can cause the memory exhaustion. 

Oh and trust me I am :P

**Comment by askmike** [comment]
> The server.js thread (as shown in windows task manager) eats up memory (I am assuming this is total memory usage as each of the backtests builds their various candle / data arrays. Even without talib or a talib / native indicator mix, this causes the memory exhaustion.

Ah yes, this could be improved: once the backtest runs all candles (trades, roundtrips) are cached and only dropped later once the backtest is done (based on [the passed request](https://github.com/askmike/gekko/blob/stable/web/routes/backtest.js#L32-L39)), we could pass a config and not cache them.

> TLDR: any backtest with any type of indicators (native, talib or a mix) can cause the memory exhaustion.

Yes but this only becomes a problem when you are stretching Gekko to its limits (16 months of 1 min candles). But even then it can be fixed by setting the following, right?

    node --max-old-space-size=8192 server.js

(all data fits in 8gb?)

**Comment by thegamecat** [comment]
Interestingly, I found the data being cached in the backtestHandler.js and I remmed the candle population (which is by far the biggest data store) as I assume this is the major data object that is being populated, assumably for the graphing. This indeed removed the memory issue.

This could be a flag that says something like; populate and return candles for UI requests only.

However, this also then leads me onto the concept of writing that data to a file for the graph rather than storing it in memory leading to way more flexibility, as you could also write the indicators to files, and then the UI builds based on the file names which can contain a unique id to the backtest, the dataset, minutes and so on. Would be bloody wonderful and follows how I am writing csvs for post backtesting review and graphing at the moment. However...again....these files can be 400MG in size so one does wonder how practical that is.

**Comment by thegamecat** [comment]
Sorry to answer the final question - typically the data size was around 1.7gig. But that's with 8 running in parallel. 

At some point I will test the talib + native indicators again.

**Comment by sheerlox** [comment]
Regarding #943 : 
```
  <-- POST /api/backtest

<--- Last few GCs --->

[21389:0x5579ffd5e6a0]    58717 ms: Scavenge 978.2 (1017.6) -> 972.8 (1022.6) MB, 51.2 / 0.0 ms  allocation failure 
[21389:0x5579ffd5e6a0]    58885 ms: Scavenge 984.2 (1023.6) -> 978.9 (1029.1) MB, 34.4 / 0.0 ms  allocation failure 
[21389:0x5579ffd5e6a0]    59169 ms: Scavenge 990.2 (1030.1) -> 984.8 (1035.1) MB, 60.4 / 0.0 ms  allocation failure 
[21389:0x5579ffd5e6a0]    60322 ms: Scavenge 996.4 (1036.1) -> 991.5 (38840.1) MB, 523.7 / 0.7 ms  allocation failure 


<--- JS stacktrace --->
Cannot get stack trace in GC.
FATAL ERROR: NewSpace::Rebalance Allocation failed - process out of memory
 1: node::Abort() [/usr/bin/nodejs]
 2: 0x5579fdb98f9e [/usr/bin/nodejs]
 3: v8::Utils::ReportOOMFailure(char const*, bool) [/usr/bin/nodejs]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/bin/nodejs]
 5: 0x5579fd1f127b [/usr/bin/nodejs]
 6: v8::internal::MarkCompactCollector::EvacuateNewSpaceAndCandidates() [/usr/bin/nodejs]
 7: v8::internal::MarkCompactCollector::CollectGarbage() [/usr/bin/nodejs]
 8: v8::internal::Heap::MarkCompact() [/usr/bin/nodejs]
 9: v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/bin/nodejs]
10: v8::internal::Heap::CollectGarbage(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*, v8::GCCallbackFlags) [/usr/bin/nodejs]
11: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/usr/bin/nodejs]
12: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/bin/nodejs]
13: 0x308cbd28ed46
  xxx POST /api/backtest 500 85s -

  Error: non-error thrown: Child process has died.
      at Object.onerror (/home/sherlock/projects/gekko/node_modules/koa/lib/context.js:105:40)
      at process._tickCallback (internal/process/next_tick.js:109:7)
```

**Comment by thegamecat** [comment]
Ok can you try running node with a memory increase flag. 

For example: 

node --max-old-space-size=8192 gekko --ui

**Comment by sheerlox** [comment]
@thegamecat already tried, changes nothing

**Comment by thegamecat** [comment]
Ok, try this:

open \core\workers\pipeline\messageHandlers\backtestHandler.js

And change lines 14-19 to:

 //     if(message.type === 'candle')
 //       candles.push(message.candle);

 //     else if(message.type === 'trade')
	 if(message.type === 'trade')
        trades.push(message.trade);


This will stop any candle history being available to the UI, but it will reduce the memory usage of Gekko. 

It may not be a perm solution if you use the UI, but it will help rule out one thing.

Also, it would be useful to know if your maths logic includes writing data to arrays that are only limited to the length of the backtest.

And finally, what candlesize are you using?

**Comment by sheerlox** [comment]
@thegamecat thx I'll try it asap :)

**Comment by askmike** [comment]
> I believe it's worth including this in the documentation as a tip (for those noobs like me anyway )

Me too! I am noting that there are a lot of common things hard to put the in the DOCS. What about a FAQ page where we could put this kind of stuff in?

**Comment by thegamecat** [comment]
I think that’s kind of what is required. I’ve been writing some stuff up in the readme of my own git so I remember all the tweaks and stuff, so yes - maybe it would be a good use of the wiki?

**Comment by askmike** [comment]
I rather keep everything in one place and add a FAQ page to the documentation!

**Comment by thegamecat** [comment]
Works for me man!
